### DISCUSSION TITLE
Is there a way to have a devShell that reuses cargoArtifacts?

### DISCUSSION THREAD

**Discussion by YPares** [ipetkov/crane#733] [Q&A]
Is there a way to make the `cargo` provided by the devShell aware of the dependencies that are already built?

Ideally I would like `cargo build` in my devShell to only build my crate, and reuse the artifacts from `craneLib.buildDepsOnly`.

**Comment by dpc** [comment]
You probably don't really want that, at least not for normal dev work. During normal dev work cargo's ability to do incremental dependency builds etc. is actually somewhat useful, while Nix can only work on the granularity of whole derivations.

My general advice: in dev shell let Nix provide dev with all the tooling that is needed, and not involve `nix` again in anything. Use `nix build` for CI.

**Reply by YPares** [reply]
I agree, but I'm working on a crate whose deps are long to build, so it would be beneficial to be able, during iterative dev, to just reuse the deps that have been built by the CI, because those deps aren't gonna change often. Especially since we are several people working on that crate, and we would all benefit from that instead of all having to rebuild all deps with cargo on our dev machines.

**Reply by ipetkov** [reply]
You can always wire up a devshell's `shellHook` to copy over/link the prebuilt artifacts. You may need to get clever to only do it when necessary (as to avoid clobbering any local changes)

**Reply by YPares** [reply]
@ipetkov Yeah, that's what I tried to do, but always ended up in cargo rebuilding everything, even if `target` was pre-populated.

**Reply by ipetkov** [reply]
If you don't have an identical environment to what was built inside the Nix sandbox you may get artifact invalidation. See https://github.com/ipetkov/crane/issues/294
